File: java/src/main/java/br/com/emendes/problems/ValidNumber.java
package br.com.emendes.problems;

/**
 * Um número válido pode ser dividido nesses componentes (em ordem):<br><br>
 * <p>
 * 1. Um número decimal ou inteiro.<br>
 * 2. (Opcional) um 'e' ou 'E', seguido por um inteiro.<br><br>
 * <p>
 * Um número decimal pode ser dividido nesses componentes (em ordem):<br><br>
 * 1. (Opcional) Um caracter sinal '-' ou '+'.<br>
 * 2. Um dos formatos a seguir.<br>
 * - Um ou mais digitos seguidos por um ponto '.'.<br>
 * - Um ou mais digitos seguidos por um ponto '.', seguido por um ou mais digitos.<br>
 * - Um ponto '.' seguido por um ou mais digitos.<br><br>
 * <p>
 * Um inteiro pode ser dividido nos seguintes componentes (em ordem):<br><br>
 * 1. (Opcional) Um caracter sinal '-' ou '+'.<br>
 * 2. Um ou mais digitos.<br><br>
 * <p>
 * Por exemplo, os seguintes são números válidos "2", "0089", "-0.1", "+3.14", "4.", "-.9", "2e10", "-90E3",
 * "3e+7", "+6e-1", "53.5e93", "-123.456e789", enquanto os seguintes não são números válidos "abc", "1a", "1e", "e3",
 * "99e2.5", "--6", "-+3", "95a54e53".<br><br>
 * <p>
 * Dado uma String s, retorne true caso s seja um número válido.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= s.length <= 20<br>
 * s consiste de letras inglesas (maiúsculas ou minúsculas), digitos (0-9), positivo '+', negativo '-', ou ponto '.'.
 */
public class ValidNumber {

  public boolean isNumber(String s) {
    boolean seenNum = false;
    boolean seenDot = false;
    boolean seenE = false;

    for (int i = 0; i < s.length(); ++i) {
      switch (s.charAt(i)) {
        case '.':
          if (seenDot || seenE)
            return false;
          seenDot = true;
          break;
        case 'e', 'E':
          if (seenE || !seenNum)
            return false;
          seenE = true;
          seenNum = false;
          break;
        case '+', '-':
          if (i > 0 && s.charAt(i - 1) != 'e' && s.charAt(i - 1) != 'E')
            return false;
          seenNum = false;
          break;
        case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
          seenNum = true;
          break;
        default:
          return false;
      }
    }

    return seenNum;
  }

}

File: javascript/problems/ValidNumber.js
/**
 * Dada uma string s, retorna se s é um número válido.
 *
 * Por exemplo, todos os seguintes são números válidos: "2", "0089", "-0.1", "+3.14", "4.", "-.9",
 * "2e10", "-90E3", "3e+7", "+6e-1", "53.5e93", "-123.456e789", enquanto os seguintes não são números válidos:
 * "abc", "1a", "1e", "e3", "99e2.5", "--6", "-+3", "95a54e53".
 *
 * Formalmente, um número válido é definido usando uma das seguintes definições:
 *
 * 1. Um número inteiro seguido por um expoente opcional.
 *
 * 2. Um número decimal seguido por um expoente opcional.
 *
 * Um número inteiro é definido com um sinal opcional '-' ou '+' seguido de dígitos.
 *
 * Um número decimal é definido com um sinal opcional '-' ou '+' seguido por uma das seguintes definições:
 *
 * 1. Dígitos seguidos de um ponto '.'.
 *
 * 2. Dígitos seguidos de um ponto '.' seguido de dígitos.
 *
 * 3. Um ponto '.' seguido de dígitos.
 *
 * Um expoente é definido com uma notação de expoente 'e' ou 'E' seguida por um número inteiro.
 *
 * Os dígitos são definidos como um ou mais dígitos.
 *
 * Restrições:
 *
 * 1 <= s.length <= 20
 *
 * s contém apenas letras (maiúsculas ou minúsculas), digitos (0-9), sinal de mais '+', sinal de menos '-',
 * ou ponto '.'
 *
 * @param {string} s
 * @return {boolean}
 */
const isNumber = function (s) {
  let seenDot = false;
  let seenE = false;
  let seenNumber = false;
  for (let index = 0; index < s.length; index++) {
    switch (s[index]) {
      case ".":
        if (seenDot || seenE) return false;
        seenDot = true;
        break;
      case "E":
      case "e":
        if (seenE || !seenNumber) return false;
        seenE = true;
        seenNumber = false;
        break;
      case "+":
      case "-":
        if (index > 0 && s[index - 1] !== "E" && s[index - 1] !== "e") return false;
        seenNumber = false;
        break;
      case "0":
      case "1":
      case "2":
      case "3":
      case "4":
      case "5":
      case "6":
      case "7":
      case "8":
      case "9":
        seenNumber = true;
        break;
      default:
        return false;
    }
  }

  return seenNumber;
};